packages/hemi-tunnel-actions/src/initiateWithdraw.ts

```diff
@@ -0,0 +1,232 @@
+
+      const { canWithdraw, reason } = await canInitiateWithdraw({
+        account,
+        amount,
+        checkBalance,
+        l1Chain,
+        l2Chain,
+      }).catch(() => ({
+        canWithdraw: false,
+        reason: 'failed to validate inputs',
+      }))
+
+      if (!canWithdraw) {
+        // reason must be defined because canWithdraw is false
+        emitter.emit('withdraw-failed-validation', reason!)
+        return
+      }
+
+      emitter.emit('pre-withdraw')
+
+      // Using @ts-expect-error fails to compile so I need to use @ts-ignore
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+      // @ts-ignore because it works on IDE, and when building on its own, but fails when compiling from the portal through next
+      const hash = await writeContract(l2WalletClient, {
+        abi: l2BridgeAbi,
+        account,
+        address: l2Bridge,
+        // See https://github.com/ethereum-optimism/ecosystem/blob/8da00d3b9044dcb58558df28bae278b613562725/packages/sdk/src/adapters/eth-bridge.ts#L178
+        args: [l2TokenAddress, amount, 0, '0x'],
+        chain: l2Chain,
+        functionName: 'withdraw',
+        value,
+      }).catch(function (error) {
+        emitter.emit('user-signing-withdraw-error', error)
+      })
+
+      if (!hash) {
+        return
+      }
+
+      emitter.emit('user-signed-withdraw', hash)
+
+      const withdrawalReceipt = await l2PublicClient
+        .waitForTransactionReceipt({
+          hash,
+        })
+        .catch(function (err) {
+          emitter.emit('withdraw-failed', err)
+        })
+      if (!withdrawalReceipt) {
+        return
+      }
+
+      emitter.emit(
+        withdrawalReceipt.status === 'success'
+          ? 'withdraw-transaction-succeeded'
+          : 'withdraw-transaction-reverted',
+        withdrawalReceipt,
+      )
+    } catch (error) {
+      emitter.emit('unexpected-error', error as Error)
+    } finally {
+      emitter.emit('withdraw-settled')
+    }
+  }
+
+export const encodeInitiateWithdraw = ({
+  amount = BigInt(0),
+  l2TokenAddress,
+}: {
+  amount: bigint | undefined
+  l2TokenAddress: Address
+}) =>
+  encodeFunctionData({
+    abi: l2BridgeAbi,
+    // See https://github.com/ethereum-optimism/ecosystem/blob/8da00d3b9044dcb58558df28bae278b613562725/packages/sdk/src/adapters/eth-bridge.ts#L178
+    args: [l2TokenAddress, amount, 0, '0x'],
+    functionName: 'withdraw',
+  })
+
+export const initiateWithdrawEth = ({
+  account,
+  amount,
+  l2PublicClient,
+  ...args
+}: Omit<InitiateWithdraw, 'checkBalance' | 'value'>) =>
+  toPromiseEvent<WithdrawEvents>(
+    runInitiateWithdraw({
+      account,
+      amount,
+      async checkBalance() {
+        const balance = await hasEthBalance({
+          account,
+          publicClient: l2PublicClient,
+        })
+        // using >= because we need to consider gas fees
+        if (amount >= balance) {
+          return 'insufficient balance'
+        }
```

Comment:
We should use ">" instead of ">=". And could we have a raw estimate of the fees so we can "amount >= balance + fee"? And I think the comment is wrong. Double check.

Reply:
> And could we have a raw estimate of the fees so we can "amount >= balance + fee"? 

We could, but I thought it would not be very easy.

Reply:
plus the wallet already does this for us 😄